### PR TITLE
Add testing for union-find

### PR DESCRIPTION
**Pull request by aprokop**
- Add union-find test
- Update `UnionFind` to have `memory_space` typedef and a `size()` function

### CODE REVIEW

**Comment by aprokop** [comment]
retest this please

**Comment by aprokop** [comment]
Hmm, one build fails with 
```
/usr/bin/ld: Dwarf Error: Invalid or unhandled FORM value: 0x25.

CMakeFiles/ArborX_Test_DetailsClusteringHelpers.exe.dir/tstUnionFind.cpp.o:
In function `void Kokkos::atomic_min<int>(int*, desul::Impl::dont_deduce_this_parameter<int const>::type)':

tstUnionFind.cpp:(.text._ZN6Kokkos10atomic_minIiEEvPT_N5desul4Impl26dont_deduce_this_parameterIKS1_E4typeE[_ZN6Kokkos10atomic_minIiEEvPT_N5desul4Impl26dont_deduce_this_parameterIKS1_E4typeE]+0x17):
undefined reference to `desul::atomic_min(int*, int, desul::MemoryOrderRelaxed, desul::MemoryScopeDevice)'
```

**Comment by aprokop** [comment]
So the undefined reference to
`desul::atomic_min(int*, int, desul::MemoryOrderRelaxed, desul::MemoryScopeDevice)'
is still a problem with the latest Kokkos `develop`.

**Comment by aprokop** [comment]
Workaround have worked.

File: test/tstUnionFind.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+#include <ArborX_DetailsUnionFind.hpp>
+#include <ArborX_DetailsUtils.hpp> // iota
+
+#include "BoostTest_CUDA_clang_workarounds.hpp"
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(UnionFind)
+
+namespace
+{
```
Comment (dalg24) [inline test/tstUnionFind.cpp:22]: Why did you add an anonymous namespace here?

File: test/tstUnionFind.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+template <typename ExecutionSpace, typename UnionFind>
+Kokkos::View<int *, Kokkos::HostSpace>
+build_representatives(ExecutionSpace const &space, UnionFind union_find)
```
Comment (dalg24) [inline test/tstUnionFind.cpp:26]: Why "build" rather than "find"?

File: test/tstUnionFind.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+} // namespace
+
+#define ARBORX_TEST_UNION_FIND_REPRESENTATIVES(space, union_find, ref)         \
+  BOOST_TEST(build_representatives(space, union_find) == ref,                  \
+             boost::test_tools::per_element());
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(union_find, DeviceType, ARBORX_DEVICE_TYPES)
+{
+  using ExecutionSpace = typename DeviceType::execution_space;
+  using MemorySpace = typename ExecutionSpace::memory_space;
```
Comment (dalg24) [inline test/tstUnionFind.cpp:82]: No that is a mistake, you need to honor the memory space specified in `DeviceType`
```suggestion
  using MemorySpace = typename DeviceType::memory_space;
```

File: test/tstUnionFind.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        representatives(i) = r;
+      });
+  // We want the representative values to not depend on a specific
+  // implementation of the union-find (e.g., not relying them being equal to
+  // the smallest index in the set), so we explicitly remap them.
+  Kokkos::parallel_for(
+      "Test::remap_representatives",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n), KOKKOS_LAMBDA(int i) {
+        representatives(i) = map2smallest(representatives(i));
+      });
```
Comment (dalg24) [inline test/tstUnionFind.cpp:59]: I struggle to convince myself that this does the right thing w/o making a temporary copy of `representative`

Comment (aprokop) [reply test/tstUnionFind.cpp:61]: This reads and then writes one value in a view. Why would you want a copy?

File: test/tstUnionFind.cpp (hunk 1 above)
Comment (aprokop) [reply test/tstUnionFind.cpp]: Removed.

File: test/tstUnionFind.cpp (hunk 2 above)
Comment (aprokop) [reply test/tstUnionFind.cpp:23]: Because I consider "find" to build a call to the union-find structure. Here, we construct an additional view outside of the union-find structure containing representatives. So I think "build" is appropriate.

File: test/tstUnionFind.cpp (hunk 4 above)
Comment (dalg24) [reply test/tstUnionFind.cpp:61]: Nevermind

File: test/tstUnionFind.cpp (hunk 2 above)
Comment (dalg24) [reply test/tstUnionFind.cpp:23]: I am not convinced but not blocking

**Review by dalg24 — APPROVED** [review]
Looks good enough.

I think that a follow up that "uncouples" finding representatives and merge operations would be good.

find_representatives test: `labels` -> `representatives`
merge test: `labels`, `merge list` -> `new labels`
